crates/forge/src/lib.rs

```diff
@@ -435,14 +408,67 @@ fn run_tests_from_crate(
     ))
 }
 
-fn run_with_fuzzing(
-    package_root: &Utf8PathBuf,
-    runner_config: &RunnerConfig,
-    runner_params: &RunnerParams,
-    runner: &SierraCasmRunner,
-    case: &TestCase,
-    args: &Vec<&ConcreteTypeId>,
-) -> Result<(TestCaseSummary, u32)> {
+fn choose_test_strategy_and_run(
+    case: Arc<TestCase>,
+    runner: Arc<SierraCasmRunner>,
+    runner_config: Arc<RunnerConfig>,
+    args: Vec<ConcreteTypeId>,
+) -> JoinHandle<Result<TestCaseSummary>> {
+    if args.is_empty() {
+        run_single_test(case, runner, runner_config)
+    } else {
+        tokio::task::spawn(async move {
+            run_with_fuzzing(runner_config.clone(), runner, case, &args).await
+        })
+    }
+}
+
+fn run_single_test(
+    case: Arc<TestCase>,
+    runner: Arc<SierraCasmRunner>,
+    runner_config: Arc<RunnerConfig>,
+) -> JoinHandle<Result<TestCaseSummary>> {
+    let exit_first = runner_config.exit_first;
+    tokio::task::spawn(async move {
+        tokio::select! {
+            () = runner_config.exit_first_cancellation_token.cancelled() => {
```

Comment:
I don't like cancellation tokens being part of runner config. I think they should be passed separately, these are more like implementation details of the task execution.

Reply:
but on the other hand cancellation is part of the runner, `exit_first` works in the context of all tasks

Reply:
It is just kind of weird to me to include it as part of the config.